### PR TITLE
[Alignment 2.0] Qualifier Ranking+Filter compiler implementation

### PR DESCRIPTION
**Pull request by michalovjan**
Dependencies:
requires https://github.com/project-ncl/pnc-api/pull/141

### Notable changes:
Addition of compiler of **Alignment Ranks**.
- Or in other words, compiler which takes a list of ranks f.e. `List.of("MILESTONE:EAP 7.2", "PRODUCT: EAP and QUALITY:RELEASED", "QUALITY:TESTED")` and transforms it into a code which can sort Artifacts' Versions (String version+qualifier-with-values). 

**AlignmentPredicate**:
- It's similar to Ranks in a way that it also contains Qualifiers but it is not a List. Qualifiers are separated by comma ',' which is logically equivalent to OR
- Compiles f.e. "QUALITY:BLACKLISTED, CATEGORY:AUTO, PRODUCT:RHSSO" into a predicate which can be used as a filter.
- More specifically, this class is used to implement allow-list and deny-list fields.

Package `alignment.ranking.tokenizer`:
- contains Tokenizer for all possible Tokens
- f.e an input of "PRODUCT:EAP and ( MILESTONE-ID:213 or MILESTONE-ID:423)" would result in
  - QVALUE(Q:PRODUCT, V:[EAP]), LOGIC(AND), LPAREN, QVALUE(Q:M-ID, V:[213]), LOGIC(OR) QVALUE(Q:M-ID, V:423), RPAREN
- tokenizer doesn't care about order, it just parses input into *Tokens
- some characters are special like ':' and parantheses. In case a value for a qualifier contains it, use quotes. F.e BUILD-CONFIG: 'yo:yo:yo'
- whitespaces are ignored where possible (between tokens and before/after ':' for qualifer-values

Package `alignment.ranking.parser`:
- contains a Parser that takes tokens from Tokenizer and constructs a syntax tree
- takes into consideration order of logic operations. AND should be resolved earlier than OR on the same level. Of course parentheses enables to override behaviour
- order of tokens matters and throws a ValidationException in case the order doesn't make sense
- to know the specifics of the algorithm and tree operations, I can explain in detail in person/meeting with examples 

Package `alignment.ranking.compiler`:
 - contains a Parser and is specific for the desired output (Predicate for filters or Comparator for rankings)
 - takes a syntax tree compiled by a Parser, traverses the tree and based on the structure of the tree, it compiles it into a single access point as filter/predicate
 - AlignmentRanking further compresses the output from the compiler to garner the interaction between ranks

### CODE REVIEW

File: src/main/java/org/jboss/pnc/common/sql/NativeQueryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        orderBys = new ArrayList<>();
+    }
+
+    public NativeQueryBuilder select(String field) {
+        selects.add(field);
+        return this;
+    }
+
+    public NativeQueryBuilder select(String table, String field) {
+        selects.add((table == null || table.isEmpty() ? "" : table + '.') + field);
```
Comment (vibe13) [inline src/main/java/org/jboss/pnc/common/sql/NativeQueryBuilder.java:31]: I would extract `table + (alias == null || alias.isEmpty() ? "" : " AS " + alias)` into a method as it's commonly used

File: src/main/java/org/jboss/pnc/common/sql/NativeQueryBuilder.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        return this;
+    }
+
+    public NativeQueryBuilder select(String table, String field, String alias) {
+        selects.add(
+                (table == null || table.isEmpty() ? "" : table + '.') + field
+                        + (alias == null || alias.isEmpty() ? "" : " AS " + alias));
```
Comment (vibe13) [inline src/main/java/org/jboss/pnc/common/sql/NativeQueryBuilder.java:38]: I would extract `(alias == null || alias.isEmpty() ? "" : " AS " + alias)` into a method as it's commonly used

File: src/main/java/org/jboss/pnc/common/sql/NativeQueryBuilder.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    public NativeQueryBuilder from(String table, String alias) {
+        froms.add(table + (alias == null || alias.isEmpty() ? "" : " AS " + alias));
+        return this;
+    }
+
+    public NativeQueryBuilder join(String joinType, String table, String alias, String onClause) {
+        joins.add(
+                joinType + " JOIN " + table + (alias == null || alias.isEmpty() ? "" : " AS " + alias)
+                        + (onClause == null || onClause.isBlank() ? "" : " ON " + onClause));
```
Comment (vibe13) [inline src/main/java/org/jboss/pnc/common/sql/NativeQueryBuilder.java:63]: I would extract `(onClause == null || onClause.isBlank() ? "" : " ON " + onClause)` into a method as it's commonly used

**Review by vibe13 — CHANGES_REQUESTED** [review]

**Comment by michalovjan** [comment]
@vibe13 rebase on top of master and addressed your comments

File: src/main/java/org/jboss/pnc/common/sql/NativeQueryBuilder.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
+package org.jboss.pnc.common.sql;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class NativeQueryBuilder {
```
[the remark this answers is not in the document]
Comment (janinko) [reply src/main/java/org/jboss/pnc/common/sql/NativeQueryBuilder.java:7]: String concatenation???

**Pull request merged**